### PR TITLE
import collector chart

### PR DESCRIPTION
**Pull request by jackkav**
**What this PR does / why we need it:**
tl;dr Introducing a new chart for Kong Brain & Immunity features.
Brain and Immunity are a set of features provided by a python API (collector), a celery queue, psql and redis

This chart currently includes instructions to create and connect kong and collector

It does not yet include any CI tests, or a pinned version of kong enterprise

The aim of this PR is to have stable pinned versions of kong chart, kong enterprise and collector which pass local integration testing.

**Special notes for your reviewer:**
- I've based the chart name "kong-collectorapi" off of kong-portalapi, but may not be ideal open to suggestions.
- If you already have a minikube environment, ensure it has the latest version of `mashape/kong-enterprise:dev-master`

**Checklist**

 - [ ] Chart Version bumped (0.1.4?)
 - [ ] Kong enterprise is pinned (?)
 - [x] Collector is pinned (1.2.1)
 - [x] Kong chart is pinned (1.3.0)
 - [x] Variables are documented in the README.md

**Testing**
With helm 3 and minikube
```
helm repo add kong https://charts.konghq.com
helm repo update

minikube start --vm-driver hyperkit --memory='6144mb' --cpus=4

helm install my-kong kong/kong --version 1.3.0 -f kong-values.yaml --set env.admin_api_uri=$(minikube ip):32001

helm install my-release . --set kongAdmin.host=my-kong-kong-admin

kubectl wait --for=condition=complete job --all && helm test my-release
```
[cd to kong-collector repo](https://github.com/Kong/kong-collector) (may need to install python packages, and ensure you have a cleaned PersistentVolumeClaims between test runs)
```
KONG_ADMIN_URL=$(minikube ip):32001  \
   COLLECTOR_URL=my-release-kong-collectorapi:5000  \
   ENDPOINT_URL=my-release-kong-collectorapi-testendpoints:6000 ./kong-setup.sh

KONG_PROXY_URL=$(minikube ip):32000  \
   KONG_PROXY_INTERNAL_URL=$(minikube ip):8000 \
   KONG_ADMIN_URL=$(minikube ip):32001  \
   COLLECTOR_URL=$(minikube ip):31555 pipenv run python integration_test.py
```
Should give the result
image

### CODE REVIEW

File: charts/collector/values.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# Set this to the Kong Admin API port
+kongAdminServicePort: 8001
+# Set this to the Kong Admin API protocol
+kongAdminProtocol: http
+containerPort: 5000
+# Set a nodePort which is available
+nodePort: 31555
+service: 
+  type: NodePort
+  port: 5000
```
Comment (hbagdi) [inline charts/collector/values.yaml]: Do you foresee a possibility of having more services in this chart?

If yes then I'd recommend to nest these properties under `collector` or whatever name of the service you want to be.

This is from the experience of Kong's chart where the default or root level properties are mostly for the proxy service and then all other services are nested under the service name.

File: charts/collector/requirements.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+dependencies:
+- name: postgresql
+  version: 8.2.1
+  repository: https://kubernetes-charts.storage.googleapis.com/
```
Comment (hbagdi) [inline charts/collector/requirements.yaml]: You should use a condition: https://github.com/Kong/charts/blob/master/charts/kong/requirements.yaml#L5

File: charts/collector/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+*/}}
+{{- define "kong-collectorapi.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "kong-collectorapi.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end -}}
+
+{{/*
+Create the name of the service account to use
+*/}}
+{{- define "kong-collectorapi.serviceAccountName" -}}
```
Comment (hbagdi) [inline charts/collector/templates/_helpers.tpl]: Where is the this being used?

**Review by hbagdi — CHANGES_REQUESTED** [review]
Looks like a step in the right direction.
I couldn't get into gritty details and reviewed from an API/user perspective.

File: charts/collector/.helmignore [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# Common backup files
+*.swp
+*.bak
+*.tmp
+*~
+# Various IDEs
+.project
+.idea/
+*.tmproj
+OWNERS
```
[the remark this answers is not in the document]
Comment (jackkav) [reply charts/collector/.helmignore]: There is a file just like this in the kong folder, am I missing something?

File: charts/collector/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+name: kong-collectorapi
```
[the remark this answers is not in the document]
Comment (jackkav) [reply charts/collector/Chart.yaml]: Collector is comprised of Brain & Immunity features

File: charts/collector/values.yaml (hunk 1 above)
Comment (jackkav) [reply charts/collector/values.yaml]: AFAIK there may be another subchart for a TSDB but it seems unlikely another service will be included.

Comment (hbagdi) [reply charts/collector/values.yaml]: I'd still recommend that you indent the these root setting under a property name of the service. That provides a nice API for the end user when they are editing this file or using `--set sname.nodePort=x`.

Comment (hbagdi) [reply charts/collector/values.yaml]: Also, because you have multiple k8s services as part of the chart installation, it is better to clarify this and set the right expectation as to what change will happen.

File: charts/kong-collectorapi/values.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+containerPort: 5000
+# Set a nodePort which is available
+nodePort: 31555
+service: 
+  type: NodePort
+  port: 5000
+
+# Enable this for running integration tests
+testendpoints: 
+  enabled: false
```
Comment (hbagdi) [inline charts/kong-collectorapi/values.yaml]: nit: Since we don't expect users to use this, we should move this to the end of the file.

File: charts/kong-collectorapi/values.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  tag: latest
+  pullPolicy: IfNotPresent
+
+
+# Set this to the Kong Admin API hostname
+kongAdminHost: my-kong-kong-admin
+# Set this to the Kong Admin API port
+kongAdminServicePort: 8001
+# Set this to the Kong Admin API protocol
+kongAdminProtocol: http
```
Comment (hbagdi) [inline charts/kong-collectorapi/values.yaml]: Since these three properties have `kongAdmin` in common, it is usually recommended to introduce a property `kongAdmin` and then nest the details under it. Something like:

```yaml
kongAdmin
  host:
  port:
  protocol:
```

File: charts/kong-collectorapi/values.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
+image:
+  # repository: kong-docker-kong-brain-immunity-base.bintray.io/kong-brain-immunity
+  # tag: latest
+  pullPolicy: IfNotPresent
+  repository: kong-docker-kong-brain-internal-builds.bintray.io/kong-brain
```
Comment (hbagdi) [inline charts/kong-collectorapi/values.yaml]: This seems wrong. We should keep line 2 as that's the one we want users to use. Same applies for the tag.

File: charts/kong-collectorapi/templates/tests/test-connection.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Pod
```
Comment (hbagdi) [inline charts/kong-collectorapi/templates/tests/test-connection.yaml]: It is usually recommended to not create Pod instances like this. A `Job` is more suitable and manageable for such use-cases.

File: charts/kong-collectorapi/kong-values.yaml [hunk 10]
```diff
@@ -0,0 +1,51 @@
+image:
+  # repository: kong-docker-kong-enterprise-edition-docker.bintray.io/kong-enterprise-edition
+  # tag: latest
+  repository: mashape/kong-enterprise
```
Comment (hbagdi) [inline charts/kong-collectorapi/kong-values.yaml:4]: This seems wrong.

File: charts/kong-collectorapi/README.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+   secret](https://github.com/Kong/charts/tree/master/charts/kong#kong-enterprise)
+
+2. [Add Kong Enterprise registry
+   secret](https://github.com/Kong/charts/tree/master/charts/kong#kong-enterprise-docker-registry-access)
+
+3. Set up Kong Enterprise with postgresql, overriding postgres host and setting
+   a port for kong manager to use the Kong Admin API
+
+```console
+$ helm install my-kong stable/kong --version 0.36.1 -f kong-values.yaml --set env.admin_api_uri=$(minikube ip):32001
```
Comment (hbagdi) [inline charts/kong-collectorapi/README.md]: While we take every step to not break anything across multiple versions of the chart, it sometimes does happen. I'd recommend to use a specific version of the chart here using `--version` flag.

Comment (jackkav) [reply charts/kong-collectorapi/README.md]: I agree `--version 0.36.1`was the last version I tested this with

File: charts/kong-collectorapi/values.yaml (hunk 8 above)
Comment (jackkav) [reply charts/kong-collectorapi/values.yaml]: Just temporarily using internal master builds until I've made a release for collector and collector-plugin is in an enterprise build.

File: charts/kong-collectorapi/kong-values.yaml (hunk 10 above)
Comment (jackkav) [reply charts/kong-collectorapi/kong-values.yaml:4]: Just temporarily using internal master builds until I've made a release for collector and collector-plugin is in an enterprise build.

**Comment by rainest** [comment]
Some notes from my initial test install, mostly focusing on values.yaml, docs, and behavior. I haven't yet performed review of the templates themselves:

**Comments**
* values.yaml should expose resource limits/requests for deployments.
* The admin token should accept a secret name/key and use a `secretKeyRef` in the deployment.
* `kong-collectorapi.wait-for-kong` doesn't fully confirm that the Collector can talk to Kong, and can allow the Collector to come up in a state where it cannot communicate with Kong. We should mimic a basic admin API request with curl or similar.
* Setup instructions should be platform-agnostic as best possible. We shouldn't assume Minikube, and should generally assume a production environment if it's not possible to write something fully agnostic. While instructions for test environments are useful (and we need to write them for the Kong chart), we want users to start from a production config and work back to a test config if needed, rather than the reverse. On the support end, it's much easier if users have relatively similar production configurations, or can be pointed to instructions that will get them one.
* We shouldn't assume NodePorts are open, as they generally aren't without intentionally creating firewall rules to open them. This gets a bit tricky since there's probably less reason to expose the collector via an Ingress or LoadBalancer than the Kong services--this and the Kong chart should probably rely on port-forwards to provide access without making other assumptions about the environment.
* We also should not specify a default NodePort in values.yaml--the option should remain, but be commented out, since the default may conflict with a user's environment. K8S defaults to choosing a random available port, which is fine for most scenarios.

**Questions**
* The chart currently tries to pull `kong-docker-kong-brain-internal-builds.bintray.io/kong-brain-test-endpoints:latest`. Is this something that users should have available/is waiting on a public repository, or is it left over from our own internal test setup?
* Is there a reason to separate the Celery containers into separate pods? Does anything need to interact with them other than the Collector, or would we expect to have different scaling requirements for them and the Collector? It looks like they could possibly be moved into a single pod and their services removed.

**Comment by jackkav** [comment]
- TODO: admin token secret :shipit: 
- I've exposed collector -> kong connectivity state at collector/status so we could use than instead of wait for kong
- I'm using minikube for now because the bintray issue keeps me from using kind, although we can strip minikube from the readme.md i'll leave it in the PR description.
- nodeports are only use for the integration test, I'm open to finding a nicer way to manage this aspect of testing.
- Agreed.

**Answers**

- test-endpoints is just an echo server with some extra functions for running integration tests.  So far it's the only thing I can use to ensure everything is running. I see no reason it couldn't be public but it's only for confirming proper installation.
- I hadn't considered the pod idea. However, currently we don't add workers to scale but we plan to, so perhaps its best to put them in one pod and separate them when scaling is an available to the user..

**Comment by jackkav** [comment]
Moved to https://github.com/Kong/kong-collector-helm
